alibi/explainers/ale.py

```diff
@@ -0,0 +1,479 @@
+
+    # pre-pend 0 for the left-most point
+    zeros = np.zeros((1, accum_p_deltas.shape[1]))
+    accum_p_deltas = np.insert(accum_p_deltas, 0, zeros, axis=0)
+
+    # mean effect, R's `ALEPlot` and `iml` version (approximation per interval)
+    ale0 = (((accum_p_deltas[:-1, :] + accum_p_deltas[1:, :]) / 2) * interval_n[1:, np.newaxis]) \
+               .sum(axis=0) / interval_n.sum()
+
+    # crude approximation (assume datapoints on interval endpoints)
+    # ale0 = accum_p_deltas.mean(axis=0)
+
+    # exact marginalisation
+    # exact_ale = accum_p_deltas[indices - 1] + ((X[:, feature] - q[indices])) / (q[indices] - q[indices - 1]) * (
+    #            accum_p_deltas[indices] - accum_p_deltas[indices - 1])
+    # ale0 = exact_ale.mean()
+
+    # center
+    ale = accum_p_deltas - ale0
+
+    return q, ale, ale0
+
+
+@no_type_check
+def plot_ale(exp: Explanation,
+             features: Union[List[Union[int, str]], str] = 'all',
+             targets: Union[List[Union[int, str]], str] = 'all',
+             n_cols: int = 3,
+             sharey: str = 'all',
+             constant: bool = False,
+             ax: Union['plt.Axes', np.ndarray, None] = None,
+             line_kw: Optional[dict] = None,
+             fig_kw: Optional[dict] = None) -> 'np.ndarray':
+    """
+    Plot ALE curves on matplotlib axes.
+
+    Parameters
+    ----------
+    exp
+        An `Explanation` object produced by a call to the `ALE.explain` method.
+    features
+        A list of features for which to plot the ALE curves. Can be a mix of integers denoting
+        feature index or strings denoting entries in `exp.feature_names`. Defaults to 'all'.
+    targets
+        A list of targets for which to plot the ALE curves. Can be a mix of integers denoting
+        target index or strings denoting entries in `exp.target_names`. Defaults to 'all'.
+    n_cols
+        Number of columns to organize the resulting plot into.
+    sharey
+        A parameter specifying whether the y-axis of the ALE curves should be on the same scale
+        for several features. Possible values are 'all', 'row', None.
+    constant
+        A parameter specifying whether the constant zeroth order effects should be added to the
+        ALE first order effects.
+    ax
+        A matplotlib axes or a numpy array of matpotlib axes to plot on.
+    line_kw
+        Keyword arguments passed to the `plt.plot` function.
+    fig_kw
+        Keyword arguments passed to the `fig.set` function.
+
+    Returns
+    -------
+    An array of matplotlib axes with the resulting ALE plots.
+
+    """
+    import matplotlib.pyplot as plt
+    from matplotlib.gridspec import GridSpec
+
+    if features == 'all':
+        features = range(0, len(exp.feature_names))
+    else:
+        for ix, f in enumerate(features):
+            if isinstance(f, str):
+                try:
+                    f = np.argwhere(exp.feature_names == f).item()
```

Comment:
is `feature_names` not a list? we could use `feature_names.index()` and avoid casting to numpy array etc

Reply:
It is actually an array (conversion to array happens in `explain`). I think I decided to go with arrays because the functionality was more consistent with all the plotting code etc. This is one for #227 imo.

Reply:
Yes, we'll decide what the best thing is then. I thought everywhere in the codebase `feature_names` was a list. Missed the line where the conversion happened!